sympy/solvers/solveset.py

```diff
@@ -998,6 +1006,203 @@ def _solveset(f, symbol, domain, _check=False):
+            modterms[0].args[0].has(symbol) and # B-> function of symbol
+            modterms[0].args[1].is_integer and # C-> to be an integer.
+            any(isinstance(term, Mod)
+            for term in list(_term_factors(f))) # free from other funcs
+            )
+
+
+def _invert_modular(modterm, rhs, n, symbol):
+    """
+    Helper function to invert modular equation.
+    ``Mod(a, m) - rhs = 0``
+
+    Generally it is inverted as (a, ImageSet(Lambda(n, m*n + rhs), S.Integers)).
+    More simplified form will be returned if possible.
+    If it is not invertible than modterm, rhs is returned.
+
+    Parameters
+    ==========
+
+    modterm, rhs : Expr
+        The modular equation to be inverted, ``modterm - rhs = 0``
+
+    symbol : Symbol
+        The variable in the equation to be inverted.
+
+    Returns
+    =======
+
+    A tuple (f_x, g_n) is being returned where f_x is modular independent function
+    of symbol and g_n being set of values f_x can have.
+
+    """
+    a, m = modterm.args
+
+    if not isinstance(a, (Dummy, Symbol, Add, Mul, Pow)):
+        return modterm, rhs
+
+    if (rhs - m).is_positive or (rhs - m) is S.Zero:
+        return symbol, EmptySet()
+
+    if a is symbol:
+        return symbol, ImageSet(Lambda(n, m*n + rhs), S.Integers)
+
+    if a.is_Add:
+        # g + h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.Zero:
+            return _invert_modular(Mod(h, m), (rhs - Mod(g, m)) % m, n, symbol)
+
+    if a.is_Mul:
+        # g*h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.One:
+            return _invert_modular(Mod(h, m), (rhs*invert(g, m)) % m, n, symbol)
+
+    if a.is_Pow:
+        # base**expo = a
+        base, expo = a.args
+        if expo.has(symbol) and not base.has(symbol):
+            # remainder -> solution independent of n of equation.
+            # m, rhs are made coprime by dividing igcd(m, rhs)
+            remainder = discrete_log(m / igcd(m, rhs), rhs, a.base)
+            # period -> coefficient of n in the solution and also referred as
+            # the least period of expo in which it is repeats itself.
+            # (a**(totient(m)) - 1) divides m. Here is link of theoram:
+            # (https://en.wikipedia.org/wiki/Euler's_theorem)
+            period = totient(m)
+            for p in divisors(period):
+                # there might a lesser period exist than totient(m).
+                if pow(a.base, p, m / igcd(m, a.base)) == 1:
+                    period = p
+                    break
+            return expo, ImageSet(Lambda(n, period*n + remainder), S.Integers)
+
+    return a, ImageSet(Lambda(n, m*n + rhs), S.Integers)
+
+
+def _solve_modular(f, symbol, domain):
+    r"""
+    Helper function for solving modular equations of type ``A - Mod(B, C) = 0``,
+    where A can or cannot be a function of symbol, B is surely a function of
+    symbol and C is an integer.
```

Comment:
Did you account for negative C?
In SymPy as in Python `%` is also defined for negative integers.
With this PR:
```
In [11]: 2 + Mod(1, -3)
Out[11]: 0

In [12]: solveset(2 + Mod(x,-3), x, Integers)
Out[12]: ∅
```
I'd say this needs some more thought or at least proper documentation.

(Note: I'm not decided on what I'd actually expect as a result, since the case of negative divisor is mainly a programming language extension of a mathematical concept.)

Reply:
Yeah I run through this lately and this and more loopholes are to be considered in upcoming commits